### PR TITLE
[ML] Set estimated memory usage for PyTorch models

### PR DESCRIPTION
**Pull request by davidkyle**
PyTorch models incorrectly report their `estimated_heap_memory_usage_bytes` as `0`. The true value is available from the model definition docs and is now set when a definition doc is PUT.

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
Why not require it on config creation again? Our up-loaders already know about the file size at the time of config creation.

**Comment by davidkyle** [comment]
> Why not require it on config creation again?

Well that would have been too simple

**Comment by davidkyle** [comment]
The PUT model definition part action has conveniently read the config already so the extra work to check the memory value and update is minimal. I like that at the trained model config is really simple as we don't required many fields, it can be done in a few lines:

```
{
    "description": "my model",
    "model_type": "pytorch",
    "inference_config": {
        "ner": {          
        }
    }
}
```

**Comment by benwtrent** [comment]
> The PUT model definition part action has conveniently read the config already so the extra work to check the memory value and update is minimal. I like that at the trained model config is really simple as we don't required many fields, it can be done in a few lines:

I have been against this estimation for a while. And now sneakily setting the estimation based off of total definitoin length seems weird.

I would much rather us explicitly set the estimation up front in creation instead of the sneaky total definition length being used as the estimation in the stack.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchStateStreamer.java [hunk 1]
```diff
@@ -70,6 +72,14 @@ public void writeStateToStream(String modelId, String index, OutputStream restor
         restorer.setSearchSize(1);
         restorer.restoreModelDefinition(doc -> writeChunk(doc, restoreStream), success -> {
             logger.debug("model [{}] state restored in [{}] documents from index [{}]", modelId, restorer.getNumDocsWritten(), index);
+            if (bytesWritten.get() != modelSize) {
+                logger.error(
+                    "model [{}] restored state size [{}] does not equal the expected model size [{}]",
+                    modelId,
+                    bytesWritten,
+                    modelSize
+                );
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchStateStreamer.java:81]: I like this. It shows user if their estimate size is off! Good thinking.

But, I think if we validate this on _start deployment, that is enough: https://github.com/elastic/elasticsearch/pull/80439

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchStateStreamer.java:81]: Actually, NVM. The validation in start deployment makes sure the total definition length is represented by the parts. 

If we move the model size estimation to be provided on configuration PUT, then it is wise to validate and log like this when the process starts up.

**Comment by davidkyle** [comment]
Closing. The useful change to log an error when the model size is not the same as the expected size is ported to #80553

**Pull request closed without merging**